PIE-2499: Create test ownership page (#2813)
buildkite/docs: pages/test_analytics/test_ownership.md

# Test ownership

Customers on the [Pro and Enterprise plans](https://buildkite.com/pricing) can assign test ownership to [Teams](/docs/team-management/permissions).

> 

Reviewer: I reckon we should do a little marketing type explanation around the problem and the solution, something like... 

```suggestion

Test ownership is critical to adopting a healthy testing culture at your organization. By defining test owners, your teams will become accountable for maintaining a fast and reliable test suite, ensuring confidence when you deploy your code.

```

This was a first pass crack at it... @mbelton-buildkite might have good words, but he's also likely slammed so we can always iterate.

Reviewer 2: Looks like this was moved to the top. All good.

> Test ownership is assigned to the tests in a specific suite per instructions in your TESTOWNER file. The team that is the default owner of a test [will be auto-assigned to fix any flakes](/docs/test-analytics/flaky-test-assignment) that occur.

Reviewer: What do you think about dropping the _per-suite_ instructions here, and we can talk about that requirement later in the docs... mainly because there's nothing stopping someone using the same file uploaded to multiple suites. 

Something like this perhaps:

```suggestion
Test ownership is managed via team assignments in a TESTOWNER file. The team that is the default owner of a test [will be automatically assigned flaky tests](/docs/test-analytics/flaky-test-assignment) to triage.
```

Author: Thanks I'll use that!

> 🚧 Buildkite test ownership is currently in private beta
> Please reach out to our support team to register for early access.

## Setting test ownership

You can upload a TESTOWNER file via this API endpoint:

```bash
curl --location 'https://analytics-api.buildkite.com/v1/test-ownerships' \
     --header "Authorization: Bearer <your-suite-api-token>" \
     -F 'file=@<your-TESTOWNER-file-location>'
```

> You might consider creating a new pipeline to automatically upload your TESTOWNER file when changes are detected.

Reviewer: 💚

## Your TESTOWNER file

> A TESTOWNER file [follows the same rules as a `.gitignore` or `CODEOWNERS` file](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners#example-of-a-codeowners-file), with one exception.

Reviewer: it would be cool if we could link to the one-exception and put it at the bottom of the doc under some kinda troubleshooting section, that way we get to keep them on the happy path.

Author: Yeah I think this section could just be moved down now as we have the example TESTOWNER file that describes the rules specifically now anyway.

# directory at the root of the test directory and any of its
# subdirectories.
/spec/packages/ packages

# The `spec/features/*` pattern will match files like
# `spec/features/application_spec.rb` but not further nested files like
# `spec/features/pipelines/application_spec.rb`.
spec/features/*  test-analytics

# In this example, pipelines owns any file in a pipelines directory
# anywhere in your test directory.
pipelines/ pipelines

# In this example, packages owns any file in the `/spec/packages`
# directory in the root of your test directory and any of its
# subdirectories.
/spec/packages/ packages

# In this example, test-analytics owns any test in a `/analytics` directory such as
# `/models/analytics`, `/features/analytics`, and `/models/organizations/analytics`. # Any tests in an `/analytics` directory will belong to team test-analytics.
**/analytics test-analytics

# In this example, pipelines owns any test in the `/spec`
# directory in the root of your test directory except for the `/spec/models/packages`
# subdirectory, as this subdirectory has its own owner packages
/spec/ pipelines
/spec/models/packages packages
```
{: codeblock-file="TESTOWNERS"}

### Teams

> A TESTOWNER uses Buildkite team slugs instead of user names. Your team slug will be your team name in snake-case. You can view your teams in your organization settings, or fetch them from our API:

Reviewer: i always get confused with case types... is snake-case `snake_case`... for example is it `test-analytics` or `test_analytics` and which one do we match? i think it's just the slug yeah?

Reviewer 3: oh it should be kebab case! Snake case is all lower case with underscores, kebab case uses lowercase with dashes

Author: AH it's kebab-case, not snake-case

Reviewer: i didn't even know it was called kebab case 🤣 i just thought it was dasherized.

- [List teams from REST API](/docs/apis/rest_api/teams)</li>
- [List teams from Graphql API](/docs/apis/graphql/schemas/object/team)</li>

> 🚧 Team permissions

> >It is important to [ensure these teams have permission to access the suite](/docs/test-analytics/permissions#manage-teams-and-permissions) the file is uploaded to, otherwise these ownership records will not be created. You can check these permissions in your organization's team settings.

Reviewer: I had to do some mental gymnastics on this sentence, talking about permissions is damn hard. 

How about?

> Teams listed in the the TESTOWNERS must have permission to access the suite before ownership records are created.

🤷🏻 i hate talking about permissions 😭 it breaks my brain.

More than 1 team may own a test, and the order of teams in your TESTOWNER file is important. The first team listed will be the default owner, and they will be auto-assigned to the test if it flakes. Any team with suite access can override this auto-assignment.

### Suites
A suite only has one active TESTOWNER file at a time. A TESTOWNER must be uploaded per suite for it to be applied.

> 

Reviewer: We could say something about multiple suites using the same TESTOWNERS file. 

```suggestion
### FAQ

#### Can multiple suites use the one TESTOWNERS
Yes, there's nothing stopping you from uploading the same file to multiple suites. 
```

> # Test ownership
> Test ownership is critical to adopting a healthy testing culture at your organization. By defining test owners, your teams will become accountable for maintaining a fast and reliable test suite, ensuring confidence when you deploy your code.

Reviewer 2: ```suggestion
# Test ownership

Test ownership is critical to adopting a healthy testing culture at your organization. By defining test owners, your teams will become accountable for maintaining a fast and reliable test suite, ensuring confidence when you deploy your code.
```

> Customers on the [Pro and Enterprise plans](https://buildkite.com/pricing) can assign test ownership to [Teams](/docs/team-management/permissions).

Reviewer 2: Test Analytics has its own 'Access controls' docs page for this, with relevant links through to the general platform-level docs about teams, currently in the Pipelines section of the Buildkite Docs, but which should eventually be pulled out into its own 'Platform' docs section.

It'd be more sensible for this link to lead to the Test Analyics page equivalent first, since it also talks about test suite-level permissions.

```suggestion
Customers on the [Pro and Enterprise plans](https://buildkite.com/pricing) can assign test ownership to [Teams](/docs/test-analytics/permissions).
```

Test ownership is managed via team assignments in a TESTOWNER file. The team that is the default owner of a test [will be automatically assigned flaky tests](/docs/test-analytics/flaky-test-assignment) to triage.

> 🚧 Buildkite test ownership is currently in private beta
> Please reach out to our support team to register for early access.

## Setting test ownership

You can upload a TESTOWNER file via this API endpoint:

```bash
curl --location 'https://analytics-api.buildkite.com/v1/test-ownerships' \
     --header "Authorization: Bearer <your-suite-api-token>" \
     -F 'file=@<your-TESTOWNER-file-location>'
```
You might consider creating a new pipeline to automatically upload your TESTOWNER file when changes are detected.

## Example TESTOWNER file

```bash
# This is a comment.
# Each line is a file pattern followed by one or more team slugs.

> # These teams will be the owners for everything in
> # the repo. While both teams own the test, the first team listed will
> # become the default owner for tests in this directory. Unless
> # a later match takes precedence, team-slug-1 will have any
> # flaky tests assigned to them by default.
> *       team-slug-1 team-slug-2

Reviewer 2: Minor tweaks...

```suggestion
# These teams will be the test owners for everything in this
# repository. While both teams own the test, the first team listed will
# become the default owner for tests in this directory. Unless
# a later match takes precedence, team-slug-1 will have any
# flaky tests assigned to them by default.
* team-slug-1 team-slug-2
```

Also, couple of queries:

1. What was the reason for the extra space between the `*` and `team-slug-1`? If this functionally does nothing, I'd suggest removing it (as I've suggested above). Otherwise, please feel free to add it back in, with an explanation as to why the extra space is important.
2. When you say "a later match takes precedence", what exactly is being matched here and by what? (Sorry for my ignorance.)

Author: The weird spacing has come from me copying over [Github's instructions for creating a CODEOWNER file](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners#example-of-a-codeowners-file) and overwriting it to match what we have. TESTOWNERSHIP uses the same rules as those set out for githubs `.gitignore` and `CODEOWNER` files, and I wanted to reuse something familiar. My assumption re the spacing is to try and show the difference between the location and the owner slugs. It's also how this [kind of file is usually laid out in real life](https://github.com/buildkite/buildkite/blob/main/.github/CODEOWNERS), but agree spacing is inconsistent e.g.
```bash
/docs docs docs-1
# as opposed to
/docs          docs docs-1
```

> When you say "a later match takes precedence", what exactly is being matched here and by what?

I think I answered this in https://github.com/buildkite/docs/pull/2813#discussion_r1641989485 by let me know if it's still unclear!

Reviewer 2: Thanks! Fixed and reworded following a more detailed discussion with you :-)

> # Order is important; the last matching pattern takes the most
> # precedence. Any test file ending with _spec.rb will be assigned
> # to the test-analytics team and not team-slug-1.
> *_spec.rb    test-analytics #This is an inline comment.

Reviewer 2: Minor tweaks ... please correct if my interpretation is incorrect ...

```suggestion
# Any test file ending with _spec.rb will be assigned
# to the test-analytics team and not team-slug-1.
# The last matching team takes precedence.
*_spec.rb test-analytics #This is an inline comment.
```

A couple of comments/questions:

1. Removed extra space (as per my previous comment).
1. Can multiple teams be specified against this file and if so, what's the behaviour in doing so?

Author: The suggested change isn't quite right; we might have touched on this in slack already, but to clarify:

> Can multiple teams be specified against this file and if so, what's the behaviour in doing so?

Yes, multiple teams can own the same test location. The team listed first will be the "default owner" of the test, meaning when a test flakes they alone will be auto-assigned to fix it. There's more detail in the section above starting with `# These teams will be the owners for everything in`

> Is this configuration meant to override the * team-name-style configurations above?

Correct, the order in this entire file is important. If our file was just:
```bash
*              team-slug-1 team-slug-2
*_spec.rb      test-analytics
```
We assign ownership of all tests matching *_spec.rb` to the `test-analytics` team. Then, excluding tests matching `*_spec.rb`, we assign all tests (`*`) to `team-slug-1` and `team-slug-2`.

Reviewer 2: Thanks! Fixed and reworded following a more detailed discussion with you :-)

> # Only Buildkite teams can be specified as test owners. Teams should

Reviewer 2: This line should probably be moved to after `# This is a comment.` above.
Then the 2nd sentence can be deleted as it's already covered by the following line.

# be identified by their slug. Teams must have
# explicit access to the suite the test belongs to. In this example,
# the pipelines team owns all .rb files.
*.rb pipelines

> # In this example, packages owns any files in the spec/packages/
> # directory at the root of the test directory and any of its
> # subdirectories.
> /spec/packages/ packages

Reviewer 2: ```suggestion
# In this example, the packages team owns any files in the
# spec/packages/ directory at the root of the test directory and
# any of its subdirectories.
/spec/packages/ packages
```

# The `spec/features/*` pattern will match files like
# `spec/features/application_spec.rb` but not further nested files like
# `spec/features/pipelines/application_spec.rb`.

> spec/features/*  test-analytics

Reviewer 2: Is there a reason for the double-space here?

Author: Answered in https://github.com/buildkite/docs/pull/2813#discussion_r1641993258

# In this example, pipelines owns any file in a pipelines directory
# anywhere in your test directory.
pipelines/ pipelines

# In this example, packages owns any file in the `/spec/packages`
# directory in the root of your test directory and any of its
# subdirectories.
/spec/packages/ packages

# In this example, test-analytics owns any test in a `/analytics` directory such as
# `/models/analytics`, `/features/analytics`, and `/models/organizations/analytics`. # Any tests in an `/analytics` directory will belong to team test-analytics.
**/analytics test-analytics

# In this example, pipelines owns any test in the `/spec`
# directory in the root of your test directory except for the `/spec/models/packages`
# subdirectory, as this subdirectory has its own owner packages
/spec/ pipelines
/spec/models/packages packages
```
{: codeblock-file="TESTOWNERS"}

## FAQs

> ### Can multiple suites use the same TESTOWNER file?
> Yes, there's nothing stopping you from uploading the same file to multiple suites. A suite only has one active TESTOWNER file at a time.

Reviewer 2: Add gaps under headings to make the source content a little more readable.

```suggestion
### Can multiple suites use the same TESTOWNER file?

Yes, there's nothing stopping you from uploading the same file to multiple suites. A suite only has one active TESTOWNER file at a time.
```

> ### Can more than one team own a test?
> More than 1 team may own a test, and the order of teams in your TESTOWNER file is important. The first team listed will be the default owner, and they will be auto-assigned to the test if it flakes. Any team with suite access can override this auto-assignment.

Reviewer 2: ```suggestion
### Can more than one team own a test?

More than 1 team may own a test, and the order of teams in your TESTOWNER file is important. The first team listed will be the default owner, and they will be auto-assigned to the test if it flakes. Any team with suite access can override this auto-assignment.
```

> ### Why aren't my teams being assigned ownership over my tests?
> A TESTOWNER file uses Buildkite team slugs instead of user names. Your team slug will be your team name in kebab-case. You can view your teams in your organization settings, or fetch them from our API:

Reviewer 2: ```suggestion
### Why aren't my teams being assigned ownership over my tests?

A TESTOWNER file uses Buildkite team slugs instead of user names. Your team slug will be your team name in kebab-case. You can view your teams in your organization settings, or fetch them from our API:
```

> - [List teams from REST API](/docs/apis/rest_api/teams)</li>
> - [List teams from GraphQL API](/docs/apis/graphql/schemas/object/team)</li>

Reviewer 2: Just a few more things I forgot about...

I don't think you need terminating `</li>` elements without a starting one. Also, this Markdown list syntax doesn't require any additional HTML syntax when it's rendered into HTML.

```suggestion
- [List teams from REST API](/docs/apis/rest_api/teams)
- [List teams from GraphQL API](/docs/apis/graphql/schemas/object/team)
```

```bash
# Example team name to slug
Pipelines => pipelines
Test Analytics => test-analytics
📦 Packages => packages
```

> Additionally, Teams listed in your TESTOWNERS file must have [permission to access the suite](/docs/test-analytics/permissions#manage-teams-and-permissions) before ownership records are created.

Reviewer 2: ```suggestion
Additionally, the teams listed in your TESTOWNERS file must have [permission to access the suite](/docs/test-analytics/permissions#manage-teams-and-permissions) before ownership records are created.
```

## Troubleshooting

A TESTOWNER file [follows the same rules as a `.gitignore` or `CODEOWNERS` file](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners#example-of-a-codeowners-file), with one exception

We do not currently support the `.gitignore` rule that allows a file path to have no corresponding team.

```bash

> # In a regular .gitignore or COEDOWNER file, the following
> # block would set test-analytics as the owner of any file in the `/specs`

Reviewer 2: ```suggestion
# In a regular .gitignore or COEDOWNER file, the following block would
# set the test-analytics team as the owner of any file in the `/specs`
```

# Test ownership

Customers on the [Pro and Enterprise plans](https://buildkite.com/pricing) can assign test ownership to [Teams](/docs/team-management/permissions).

Test ownership is assigned to the tests in a specific suite per instructions in your TESTOWNER file. When a test flakes it will be automatically assigned to the Buildkite team that owns it.

> 📘 Buildkite test ownership is currently in private beta
> Please reach out to our support team to register for early access.

## Your TESTOWNER file

### Formatting
A TESTOWNER file [follows the same rules as a `.gitignore` or `CODEOWNERS` file](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners#example-of-a-codeowners-file), with a few exceptions.

Order your test paths from least specific to most specific to ensure ownership records are not incorrectly overridden.

```bash
# TESTOWNERS
* everyone
/spec/models pipelines
/spec/models/examples/* test-analytics packages
/spec/models/**/test_spec.rb platform
```

> We currently do not support

Author: Lol oops \*insert github blank teams rules\*
